overseer/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+		candidate_backing: Box<dyn Subsystem<CandidateBackingSubsystemMessage> + Send>,
+		mut s: S,
+	) -> (Self, OverseerHandler) {
+		let (events_tx, events_rx) = mpsc::channel(CHANNEL_CAPACITY);
+
+		let handler = OverseerHandler {
+			events_tx: events_tx.clone(),
+		};
+
+		let mut running_subsystems_rx = StreamUnordered::new();
+		let mut running_subsystems = FuturesUnordered::new();
+
+		let validation_subsystem = Self::spawn(
+			&mut s,
+			&mut running_subsystems,
+			&mut running_subsystems_rx,
+			validation,
+		);
+
+		let candidate_backing_subsystem = Self::spawn(
+			&mut s,
+			&mut running_subsystems,
+			&mut running_subsystems_rx,
+			candidate_backing,
+		);
+
+		let this = Self {
+			validation_subsystem,
+			candidate_backing_subsystem,
+			s,
+			running_subsystems,
+			running_subsystems_rx,
+			events_rx,
+		};
+
+		(this, handler)
+	}
+
+	// Stop the overseer.
+	async fn stop(mut self) {
+		if let Some(ref mut s) = self.validation_subsystem.instance {
+			let _ = s.tx.send(FromOverseer::Signal(OverseerSignal::StopWork)).await;
+		}
+
+		if let Some(ref mut s) = self.candidate_backing_subsystem.instance {
+			let _ = s.tx.send(FromOverseer::Signal(OverseerSignal::StopWork)).await;
+		}
+
+		let mut stop_delay = Delay::new(Duration::from_secs(STOP_DELAY)).fuse();
+
+		loop {
+			select! {
+				_ = self.running_subsystems.next() => {
+					if self.running_subsystems.is_empty() {
+						break;
+					}
+				},
+				_ = stop_delay => break,
+				complete => break,
+			}
+		}
+	}
+
+	/// Run the `Overseer`.
+	pub async fn run(mut self) {
+		loop {
+			while let Poll::Ready(Some(msg)) = poll!(&mut self.events_rx.next()) {
+				match msg {
+					Event::MsgToSubsystem(msg) => {
+						self.route_message(msg).await;
+					}
+					Event::Stop => return self.stop().await,
+					_ => ()
+				}
+			}
+
+			while let Poll::Ready(Some((StreamYield::Item(msg), _))) = poll!(
+				&mut self.running_subsystems_rx.next()
+			) {
+				match msg {
+					ToOverseer::SubsystemMessage(msg) => self.route_message(msg).await,
+					ToOverseer::SpawnJob { s, res } => {
+						let s = self.spawn_job(s);
+
+						let _ = res.send(s);
+					}
+				}
+			}
+
+			// Some subsystem exited? It's time to panic.
+			if let Poll::Ready(Some(finished)) = poll!(self.running_subsystems.next()) {
+				panic!("Subsystem finished unexpectedly {:?}", finished);
+			}
+
+			// Looks like nothing is left to be polled, let's take a break.
+			pending!();
+		}
+	}
+
+	async fn route_message(&mut self, msg: AllMessages) {
+		match msg {
+			AllMessages::Validation(msg) => {
+				if let Some(ref mut s) = self.validation_subsystem.instance {
+					let _= s.tx.send(FromOverseer::Communication { msg }).await;
+				}
+			}
+			AllMessages::CandidateBacking(msg) => {
+				if let Some(ref mut s) = self.candidate_backing_subsystem.instance {
+					let _ = s.tx.send(FromOverseer::Communication { msg }).await;
+				}
+			}
+		}
+	}
+
+	fn spawn<M: Debug>(
```

Comment:
I prefer functions not taking `self` to be free functions rather than inherent impls.

Reply:
done